### ISSUE TITLE
[BUG] right_inverse with 1d layout results "#2700-D: attempt to access expired storage" on constexpr return value.

### ISSUE THREAD

**Issue by cloudhan** [NVIDIA/cutlass#1285]
**Describe the bug**

**Steps/Code to reproduce bug**
```cuda
#include "cute/tensor.hpp"
using namespace cute;
__global__ void kernel() {
  constexpr auto weird = right_inverse(make_layout(_2{}, _1{}));
  print(weird);
}

int main() {
  kernel<<<1, 1>>>();
  cudaDeviceSynchronize();
}
```

compile the code cause 
```
error: expression must have a constant value
include/cute/layout.hpp(1143): note #2700-D: attempt to access expired storage
include/cute/algorithm/tuple_algorithms.hpp(114): note #2693-D: called from:
include/cute/algorithm/tuple_algorithms.hpp(248): note #2693-D: called from:
include/cute/layout.hpp(1143): note #2693-D: called from:
```

remove the `constexpr` however, the program compiles smoothly and prints `_2:_1` 

**Expected behavior**
It compiles.

**Environment details (please complete the following information):**
  reproduced with cuda 11.8 and 12.1

**Additional context**

a75b4ac483166189a45290783cb0a18af5ff0ea5

**Comment by hwu36** [comment]
@thakkarV

**Comment by thakkarV** [comment]
@ccecka

**Comment by ccecka** [comment]
Weird one, but I think I know where this is coming from. I'll include a fix in the next CuTe review

**Comment by mhoemmen** [comment]
I do get a lot of warnings with MSVC about CuTe functions returning temporary storage.  I suspect MSVC doesn't follow the reference forwarding rules of other compilers

**Comment by cloudhan** [comment]
comment to avoid auto-close

**Comment by mhoemmen** [comment]
@ccecka has this been fixed?

**Comment by ccecka** [comment]
Doesn't appear to be, no.

**Comment by leimao** [comment]
@ccecka Still no fix? I got a problem probably relevant to this when I tried to make `make_tiled_copy` constexpr.

**Comment by thakkarV** [comment]
question, are you using the build flags generated by the CUTLASS CMake?

**Comment by leimao** [comment]
> question, are you using the build flags generated by the CUTLASS CMake?

Thanks @thakkarV. I am new to this. Could you please point me to those build flags?

I was simply adding `constexpr`s to the functions in this [file](https://github.com/NVIDIA/cutlass/blob/bf9da7b76c766d7ee7d536afc77880a4ef1f1156/include/cute/atom/copy_atom.hpp) and I was using my own CMake script for building. I might have missed something critical.

**Comment by leimao** [comment]
More specifically, is there any problem to make this [`make_tiled_copy`](https://github.com/NVIDIA/cutlass/blob/bf9da7b76c766d7ee7d536afc77880a4ef1f1156/include/cute/atom/copy_atom.hpp#L475) function `constexpr`?

**Comment by thakkarV** [comment]
we dont support custom cmake builds. please use the one that is in our repo.

**Comment by leimao** [comment]
> we dont support custom cmake configs. please use the one that is in our repo.

But... Eventually the user will write their own scripts for building the program, right? Could you please point me to the key CMake files where those build flags were set?

**Comment by thakkarV** [comment]
No, that's not the case. if you are importing CUTLASS through your own project, then you can https://stackoverflow.com/questions/56561591/link-one-project-to-another-cmake
